project/src/com/google/daggerqueryui/scripts/binding_graph.js

```diff
@@ -118,9 +123,29 @@ const bindingGraph = (function() {
     const sourceId = addNode(source);
     const targetId = addNode(target);
 
+    // Checks if an edge already exists.
+    if (isEdgeExists(sourceId, targetId)) {
+      return;
+    }
+
     edges.add({from: sourceId, to: targetId});
   }
 
+  /**
+   * Checks if an edge exists in the graph or not.
+   *
+   * @param {number} sourceId
+   * @param {number} targetId
+   * @return {boolean}
+   */
+  function isEdgeExists(sourceId, targetId) {
+    return edges.get({
+      filter: function (edge) {
```

Comment:
Could be simplified into a lambda using the js arror notation if you want `filter: edge => edge.from === sourceId && edge.to === targetId`

Reply:
Definitely it looks better, thanks! [Done](https://github.com/googleinterns/dagger-query/pull/45/commits/ef15c524f5c8194d5ddb68958b11bed1e01642fe)